### PR TITLE
Unit tests for ApplicationErrorScanner

### PR DESCRIPTION
**Pull request by adioss**
Hello,
Let's try to a PR for ApplicationErrorScanner test.
1. first commit on this repo
2. not easy to do parametrized tests with JUnit... so only one test for Regex evidence...
3. not signed/found any Contributor Licensing Agreement for OWASP. Can you help me to do that?

I hope it can help.

Adrien

### CODE REVIEW

**Comment by thc202** [comment]
I'd ask you to rename the class to end with `UnitTest` (to be run with Travis) but it will most likely fail because of missing `application_errors.xml`.
We should address that...

Regarding your comments:
2. It could be used a Parameterized runner or do it in the test "manually" (list of evidences and When / Then each one in a loop);
3. ZAP does not have a CLA (yet?), contributions are not directly related to OWASP (AFAIK).

**Comment by adioss** [comment]
My way to correct Travis (modifying the build.xml and/or pre test stuff before junit tag...) is, let's say, not that elegant... do you think we can do it better?
For the test of application_errors.xml:
* in my opinion, only regex are interesting. I prefer the loop solution: i've tried yesterday with Parameterized but all tests need inside your test class should be parameterized and/so other tests won't run (or I found a solution with inner classes, that is not cool at all, or i can create 2 test classes or i don't know how to do it...)
* or we can test all application_errors.xml file (string + regex) + check that when it's modified, all patterns are tested (do a test that count Pattern tag occurrences and failed if new appended and not tested) but maybe it's a little bit overkill...?

**Comment by thc202** [comment]
Could you rebase the pull request? The test target was changed to copy add-on resources prior running the tests (which should address this problem).

Good questions, we could have other test class (Parameterized?) to verify all regexes/strings of the "default" file and use another file for unit testing (which would have less strings/regexes to test and would not have to be changed). This would require changing the scanner to allow set the file to be used.

**Comment by kingthorin** [comment]
@adioss are you able to address @thc202's request?

**Comment by adioss** [comment]
I'm in hollidays = no computer... But at the end of next week, yep, no pb.

**Comment by thc202** [comment]
Not sure worth testing the current/default file, using a test file should be enough.

**Comment by adioss** [comment]
Rebased and after I tried something: I created ApplicationErrorScannerRegexMatcherUnitTest to test default application_errors.xml. As I said previously, in my opinion, testing string type pattern is not interesting (we don't want to test that ContentMatcher/BoyerMooreMatcher work correctly but that content of the file is correct and for string, it's quite straightforward) so I didn't test them but all the regex definitions.
In ApplicationErrorScannerUnitTest, I kept the test to check that an evidence can be detected with the matcher.
WDYT?

**Comment by adioss** [comment]
Up

File: test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerRegexMatcherUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                {"Servlet Error</title>"},
+        });
+    }
+
+    @SuppressWarnings("WeakerAccess")
+    @Parameter
+    public String input;
+
+    @Test
+    public void shouldRaiseAlertIfResponseCodeIsInternalError() {
```
Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerRegexMatcherUnitTest.java]: The name should be changed.

File: test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+package org.zaproxy.zap.extension.pscanrules;
+
+import org.junit.Test;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpMalformedHeaderException;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+import static java.lang.String.format;
+import static org.apache.commons.httpclient.HttpStatus.*;
```
Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerUnitTest.java]: Could/should be used `HttpStatusCode` instead.

File: test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        // Given
+        HttpMessage msg = new HttpMessage();
+        msg.setRequestHeader(REQUEST_HEADER);
+        msg.setResponseHeader(createResponseHeader(SC_INTERNAL_SERVER_ERROR));
+        // When
+        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+        // Then
+        assertThat(alertsRaised.size(), equalTo(1));
+        Alert result = alertsRaised.get(0);
+        assertThat(result.getEvidence(), equalTo("HTTP 500 Internal server error"));
```
Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerUnitTest.java]: This is a bug in the scanner, that evidence is not in the response.

**Comment by thc202** [comment]
That sounds good to me. Could you rebase and fixup the commits?

File: test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerUnitTest.java (hunk 2 above)
Comment (adioss) [inline test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerUnitTest.java]: sure

Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerUnitTest.java]: Yes, worth changing that too.

File: test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerRegexMatcherUnitTest.java (hunk 1 above)
Comment (adioss) [inline test/org/zaproxy/zap/extension/pscanrules/ApplicationErrorScannerRegexMatcherUnitTest.java]: done

**Comment by adioss** [comment]
Rebased and commits fixed

**Comment by thc202** [comment]
Looks good to me, thank you!

**Pull request merged**